src/epgsql_sock.erl

```diff
@@ -490,7 +490,17 @@ send_multi(#state{mod = Mod, sock = Sock}, List) ->
 do_send(gen_tcp, Sock, Bin) ->
     gen_tcp_send(Sock, Bin);
 do_send(ssl, Sock, Bin) ->
-    ssl:send(Sock, Bin).
+    case ssl:send(Sock, Bin) of
+        ok ->
+            ok;
+        {error, Reason} = Error ->
+            %% In case ssl:send/2 on an active socket fails, the controlling process doesn't get
+            %% notified, see https://github.com/erlang/otp/issues/7431 for reference.
+            %%
+            %% As a workaround, send an ssl_error message on our own.
+            self() ! {ssl_error, Sock, Reason},
```

Comment:
1. Hmm, it will definitely work. But given that `epgsql` tends to grow quite large message queue from time to time, wouldn't it be better to stop stright away (eg, using `throw`)?
2. since now we have started explicitly handling `ssl:send` errors, it might be nice to do that for `gen_tcp` as well. Maybe a separate PR..
3. Would it be possible to create some test that covers this?

Reply:
1. I'm not worried that much about the large message queue. I believe that would typically represent data incoming from the DB, correct? So stopping right away would mean losing that unprocessed data, while sending an `ssl_error` message would allow the socket to process it. This is how I would expect `ssl` to work, and it also seems to be the same direction `gen_tcp` is moving, [see the note](https://www.erlang.org/doc/man/gen_tcp.html) on _Non blocking send_ behaviour:

> An background send will detect a 'remote close' and (the inet driver will) mark the socket as 'closed'. [...]
>
> This behaviour is **not** replicated by the socket implementation. A send operation will detect a remote close and immediately return this to the caller, but do nothing else. A reader will therefore be able to extract any data from the OS buffers. If the socket is set to active to not false, the data will be received as expected (`{tcp, ...}` and then a closed message (`{tcp_closed, ...}`) will be received (not an error).

2. I believe with `gen_tcp` a `{tcp_closed, ...}` message will be sent by OTP. I don't have an idea how to get into a similar situation with that protocol. For example tracing the underlying port reveals that even when `ssl:send/2` fails, the `{tcp_closed, ...}` message is sent as expected. The bug is that it won't get translated to an `{ssl_closed, ...}` message by the processes started in the `ssl` layer. So I strongly believe I only need to "patch" `ssl:send/2` to behave similar to `gen_tcp`, and there's nothing to do with the `gen_tcp` case.

3. Yes, I think so. I'll try to come up with something.